Add $modx->deprecated method to log usage of deprecated methods

Mark-H opened the pull request:
### What does it do?
This PR adds a new method to the modX object, `deprecated`, which is to be used to log usage of deprecated methods across the core. Each unique message (deprecated method + caller) is only logged once per request. A setting is added that allows disabling the logging. 

I've also done a search for `@deprecated` docblocks, and added appropriate messages in various locations. Some code was deprecated way back in 2.1, so I look forward to seeing that removed in 3.0. 

### Why is it needed?
As we get ready to cleanup some code with 3.0, it's important users can be sure their site can be safely upgraded. Old or custom extras may use functionality that is no longer supported and will be removed in 3.0, which will require manual intervention. 

The `deprecated` method will allow a site to function normally with the deprecated methods still available in 2.7, but thanks to the logging the site owner will have more information on what should be taken care of first. 

### Related issue(s)/PR(s)
#13199, which mentions some deprecated methods for 3.0.

Mark-H committed:
Add $modx->deprecated method to log usage of deprecated methods

This is to be used as a standardised way to mark functionality as deprecated, by calling it from within the deprecated function with at least the version it was marked as deprecated in. Additional information like the calling function is automatically added, to ease identifying code that needs to be updated.

If the deprecated function cannot be automatically determined properly (e.g. the deprecated function is a snippet, which would show up as "include" in "modScript" being deprecated, or if a deprecation pertains about the usage of a full object) the third param can be provided to manually name the deprecated function.

Mark-H committed:
Flag some previously deprecated code as such with the new deprecated method

Jako commented on core/model/modx/modrequest.class.php:
```diff
@@ -526,6 +526,12 @@ public function checkPublishStatus() {
+        // core/model/modx/processors/system/config.js.php, which runs on each
+        // manager page, so leaving the deprecated notice commented out until
+        // that is resolved.
+        // $this->modx->deprecated('2.3.0', 'Support for modAction has been replaced with routing based on a namespace and action name.', 'modAction support');
```
Should that be executed?

Mark-H replied:
That's literally explained in the lines above it. 

It's a deprecated method (per the docblock), that should not be used, but every single manager request currently triggers it.

JoshuaLuckers replied:
Add a @todo to the comment so it’s easier to detect.

Jako replied:
Sorry. I only have seen the commented out line and forgot to read the comment.

Mark-H replied:
I've removed all of the comments. The method is already tagged with `@deprecated`. When `modAction` is removed from 3.x this method should also be removed.

Jako reviewed (dismissed):
I don't have a clue, why this line is commented out.

Mark-H committed:
Remove comments for deprecated modRequest->getAllActionIDs method that we cannot yet deprecate

It's still in use in the core, so would cause an entry to the log on each manager request, which needs to be resolved first in a spearate PR that handles the removal of modAction.

Mark-H committed:
Short array syntax is not supported on 5.3

Looking forward to the day we can use short syntax all over the core. :D

opengeek reviewed (approved):
This may not be the ideal solution, but it will work for now and provides an abstraction for better solutions moving forward.
